zjit/src/codegen.rs

```diff
@@ -281,11 +281,16 @@ fn gen_function(cb: &mut CodeBlock, iseq: IseqPtr, function: &Function) -> Optio
         // Compile all instructions
         for &insn_id in block.insns() {
             let insn = function.find(insn_id);
-            if gen_insn(cb, &mut jit, &mut asm, function, insn_id, &insn).is_none() {
-                debug!("Failed to compile insn: {insn_id} {insn}");
-                incr_counter!(failed_gen_insn);
-                return None;
-            }
+            let Err(last_snapshot) = gen_insn(cb, &mut jit, &mut asm, function, insn_id, &insn) else {
+                // It's fine; we generated the instruction
+                continue;
+            };
+            debug!("ZJIT: gen_function: Failed to compile insn: {insn_id} {insn}. Generating side-exit.");
+            incr_counter!(failed_gen_insn);
+            gen_side_exit(&mut jit, &mut asm, &SideExitReason::UnhandledInstruction(insn_id), &function.frame_state(last_snapshot));
+            // Don't bother generating code after a side-exit. We won't run it.
+            // TODO(max): Generate ud2 or equivalent.
+            break;
```

Comment:
I'm a fan of small indentation and doing early returns (it's actually `continue` here) for errors, but with this code, it feels like `continue;` was the error and the other was the main body of the loop, especially when you had to explain "it's fine".

Maybe that's just me, but it feels less confusing to keep the previous structure:

```suggestion
            if let Err(last_snapshot) = gen_insn(cb, &mut jit, &mut asm, function, insn_id, &insn) {
                debug!("ZJIT: gen_function: Failed to compile insn: {insn_id} {insn}. Generating side-exit.");
                incr_counter!(failed_gen_insn);
                gen_side_exit(&mut jit, &mut asm, &SideExitReason::UnhandledInstruction(insn_id), &function.frame_state(last_snapshot));
                // Don't bother generating code after a side-exit. We won't run it.
                // TODO(max): Generate ud2 or equivalent.
                break;
            }
```